### PR TITLE
Implement multi-encoder with hierarchical attention (dirha_wsj recipe)

### PR DESCRIPTION
**Pull request by ruizhilijhu**
I have implemented architecture of multiple parallel encoders with hierarchical attention fusion based on our ICASSP paper. https://arxiv.org/pdf/1811.04903.pdf.  [pytorch backend only]

Details:

- DIRHA_WSJ recipe. A dataset with multiple micrphone array. 

1. asr1: single array recipe 
2. asr1_mularr: multiple array recipe, with pretrain.

- Implemented multiple encoder setup with test module. 

1.  espnet/asr/pytorch_backend/asr_mulenc.py
2. espnet/nets/pytorch_backend/e2e_asr_mulenc.py
3. test/test_e2e_asr_mulenc.py

- Implemented hierarchical attention network (han).

### CODE REVIEW

**Comment by sw005320** [comment]
Thanks.
I may change this PR to v.0.6.0 from master as it seems to include several major changes.

**Comment by sw005320** [comment]
FYI, you can perform several tests in your local by following https://github.com/espnet/espnet/blob/master/CONTRIBUTING.md#unit-testing

**Comment by ruizhilijhu** [comment]
Thanks @sw005320 , this is very helpful! Sorry for the endless notification email.

**Comment by sw005320** [comment]
@ruizhilijhu, we usually want to have smaller PRs to make our reviews easier and also quickly merge the piece of studies.
How about splitting a PR for the pure dirha_wsj recipe and multi-array one? 
The multi-array one is still very large and I want to think of more splits but we can think of it later.

**Comment by ruizhilijhu** [comment]
@sw005320 , sure thing. That's a good suggestion. I will try to split it into smaller PRs.

File: espnet/nets/pytorch_backend/rnn/decoders.py [hunk 1]
```diff
@@ -232,45 +260,68 @@ def recognize_beam(self, h, lpz, recog_args, char_list, rnnlm=None, strm_idx=0):
         """beam search implementation
```
Comment (ShigekiKarita) [inline espnet/nets/pytorch_backend/rnn/decoders.py]: Can you also update `ScorerInterface` methods (init_state, score) as same as this `recognize_beam`? We are moving into the interface at this time.

**Comment by ShigekiKarita** [comment]
I think there are too many `getattr` in this PR. I feel this PR is for v.0.6.0 because v.0.6.0 does not care about the backward compatibility.

File: espnet/nets/pytorch_backend/e2e_asr_mulenc.py [hunk 2]
```diff
@@ -0,0 +1,612 @@
+#!/usr/bin/env python3
+
+"""Define e2e module for multi-encoder network."""
```
Comment (ShigekiKarita) [inline espnet/nets/pytorch_backend/e2e_asr_mulenc.py]: Put your paper URL in this module docstring and E2E docstring.

File: espnet/nets/pytorch_backend/e2e_asr_mulenc.py [hunk 3]
```diff
@@ -0,0 +1,612 @@
...
+from espnet.nets.pytorch_backend.rnn.attentions import att_for
+from espnet.nets.pytorch_backend.rnn.decoders import decoder_for
+from espnet.nets.pytorch_backend.rnn.encoders import encoder_for
+from espnet.nets.scorers.ctc import CTCPrefixScorer
+from espnet.utils.cli_utils import strtobool
+
+CTC_LOSS_THRESHOLD = 10000
+
+
+class Reporter(chainer.Chain):
```
Comment (ShigekiKarita) [inline espnet/nets/pytorch_backend/e2e_asr_mulenc.py]: If this reporter is completely the same to the one in `e2e_asr.py`, please import it.

File: espnet/bin/asr_recog.py [hunk 4]
```diff
@@ -173,14 +177,18 @@ def main(args):
             from espnet.asr.chainer_backend.asr import recog
             recog(args)
         elif args.backend == "pytorch":
-            # Experimental API that supports custom LMs
-            if args.api == "v2":
-                from espnet.asr.pytorch_backend.recog import recog_v2
-                recog_v2(args)
+            if args.num_encs == 1:
```
Comment (ShigekiKarita) [inline espnet/bin/asr_recog.py]: If you wil skip implementing V2 AP for multi enc in this PR, `raise NotImplementedError(f"--num-encs {args.num_encs} > 1 is not supported in --api v2")` at this time.

File: espnet/asr/pytorch_backend/asr_mulenc.py [hunk 5]
```diff
@@ -0,0 +1,536 @@
+#!/usr/bin/env python3
+
+"""Define asr module for multi-encoder network."""
```
Comment (ShigekiKarita) [inline espnet/asr/pytorch_backend/asr_mulenc.py]: Is it possible to unify `asr_multienc.py` features into `asr.py`? I think diff is not so large.

File: espnet/nets/pytorch_backend/rnn/decoders.py (hunk 1 above)
Comment (ruizhilijhu) [reply espnet/nets/pytorch_backend/rnn/decoders.py]: Sure, thanks! I will do that

File: espnet/nets/pytorch_backend/e2e_asr_mulenc.py (hunk 2 above)
Comment (ruizhilijhu) [reply espnet/nets/pytorch_backend/e2e_asr_mulenc.py]: Thanks for the suggestion! Will do

File: espnet/bin/asr_recog.py (hunk 4 above)
Comment (ruizhilijhu) [reply espnet/bin/asr_recog.py]: I will change it!

File: espnet/asr/pytorch_backend/asr_mulenc.py (hunk 5 above)
Comment (ruizhilijhu) [reply espnet/asr/pytorch_backend/asr_mulenc.py]: I could try to unify both togther!

**Comment by ruizhilijhu** [comment]
@ShigekiKarita Thanks for reviewing the code. As @sw005320 suggested, this PR is too big to review. I am splitting this PR into smaller PRs so that reviewers could look at the code more clearly. I will close this thread, and do several small PRs. Also, I will take into consideration all your suggestions here.

**Pull request closed without merging**